### PR TITLE
[Issue-329] Add support for Flink 1.10

### LINKED ISSUE #329: Update for Flink 1.10 release
**Problem description**
The code base need to adapt the coming Flink `1.10.0`

**Problem location**
Code, test and branching strategy

**Suggestions for an improvement**
First we should resolve all the compile/build/test errors using Flink version `1.10.0`. We should fork a `r0.8-flink1.9` branch first to track Flink 1.9, then raise a PR to bump up the Flink version in master.

### PR DESCRIPTION
**Pull request by AlexanderZhao1**
**Change log description**

* Updated Flink version to `1.10.0` and add Blink planner dependencies
* Refactoring the way of creating TableSource and TableSink , introducing `ConnectorCatalogTable` api
* Delete deprecated `SchematicDescriptor<TestTableDescriptor>` api
* Replace the deprecated `RecordWriters` with `RecordWriterDelegate` 
* Refactoring the testcases related with above updates.
* Change `compare` function implementation in FlinkPravegaTableITCase

**Purpose of the change**
Fixes #329 

**What the code does**
Updates Flink version support and ensures tests are passing
But we still have one test case unsolved, and it may be related to the
changing of the way of creating table source or sink in Flink 1.10. And
I already have opened another issue to fix this testcase #341

**How to verify it**
`./gradlew clean build` should pass

### CODE REVIEW

File: documentation/src/docs/table-api.md [hunk 1]
```diff
@@ -80,18 +80,36 @@ final Map<String, String> propertiesMap = DescriptorProperties.toJavaMap(desc);
 final TableSource<?> source = TableFactoryService.find(BatchTableSourceFactory.class, propertiesMap)
         .createBatchTableSource(propertiesMap);
 
-tableEnv.registerTableSource("MyTableRow", source);
+//tableEnv.registerTableSource("MyTableRow", source);
```
Comment (crazyzhou) [inline documentation/src/docs/table-api.md]: We need this method to register a table though it's deprecated.

File: src/main/java/io/pravega/connectors/flink/FlinkPravegaTableFactoryBase.java [hunk 2]
```diff
@@ -40,6 +40,10 @@
 import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_TYPE;
 import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_PROPERTY_VERSION;
 import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_VERSION;
+import static org.apache.flink.table.descriptors.DescriptorProperties.WATERMARK;
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/FlinkPravegaTableFactoryBase.java]: These lines are for tests and no longer needed. We can remove them

File: src/main/java/io/pravega/connectors/flink/FlinkPravegaTableSink.java [hunk 3]
```diff
@@ -74,6 +76,16 @@ public void emitDataStream(DataStream<Row> dataStream) {
         dataStream.addSink(writer);
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/FlinkPravegaTableSink.java]: We can have empty impl according to Flink document of `TableSink`.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableFactoryTest.java [hunk 4]
```diff
@@ -38,9 +39,9 @@
 public class FlinkPravegaTableFactoryTest {
 
     final static Schema SCHEMA = new Schema()
-                                            .field("name", Types.STRING )
-                                            .field("age", Types.INT );
-    final static Json JSON = new Json().failOnMissingField(false) .deriveSchema();
+                                            .field("name", DataTypes.STRING())
+                                            .field("age", DataTypes.INT());
+    final static Json JSON = new Json().failOnMissingField(false).deriveSchema();
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableFactoryTest.java]: Derive schema is automatically done and deprecated, please remove.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 5]
```diff
@@ -28,20 +28,25 @@
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
 import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
-import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.api.EnvironmentSettings;
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: Sort the dependencies

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 6]
```diff
@@ -229,7 +242,8 @@ public void testTableSourceUsingDescriptor() throws Exception {
         execEnvWrite.execute("PopulateRowData");
 
         testTableSourceStreamingDescriptor(stream, pravegaConfig);
-        testTableSourceBatchDescriptor(stream, pravegaConfig);
+        //TODO fix the Timestamp unmatch bug
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: Please write down the issue number and Flink JIRA number for further tracking

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 7]
```diff
@@ -312,8 +332,13 @@ private void testTableSourceBatchDescriptor(Stream stream, PravegaConfig pravega
         final Map<String, String> propertiesMap = desc.toProperties();
         final TableSource<?> source = TableFactoryService.find(BatchTableSourceFactory.class, propertiesMap)
                 .createBatchTableSource(propertiesMap);
+        ConnectorCatalogTable<?, ?> connectorCatalogSourceTable = ConnectorCatalogTable.source(source, false);
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: The old API `registerTableSource` API can still work here, it is correct. We can revert the change here, and found the best way for Blink planner to create the table after fixing #356  . We can revert the document change for this, too

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableSourceTest.java [hunk 8]
```diff
@@ -197,13 +196,19 @@ public void testTableSourceDescriptorWithWatermark() {
...
 
+        final TableSchema tableSchema = TableSchema.builder()
+                .field(cityName, DataTypes.STRING())
+                .field(total, DataTypes.INT())
+                .field(eventTime, DataTypes.TIMESTAMP(3))
+                .build();
+
         final TestTableDescriptor testDesc = new TestTableDescriptor(pravega)
-                .withFormat(new Json().failOnMissingField(false) .deriveSchema())
+                .withFormat(new Json().failOnMissingField(false).deriveSchema())
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableSourceTest.java]: Same as above

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaWriterTest.java [hunk 9]
```diff
@@ -21,19 +21,21 @@
...
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
-import java.util.ArrayList;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 import java.util.UUID;
+import java.util.Optional;
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaWriterTest.java]: Sort the dependencies here.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableSourceTest.java [hunk 10]
```diff
@@ -256,7 +261,7 @@ protected TestableBuilder builder() {
     /**
      * Test Table descriptor wrapper
      */
-    static class TestTableDescriptor extends TableDescriptor implements SchematicDescriptor<TestTableDescriptor> {
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableSourceTest.java]: This API is removed in https://github.com/apache/flink/pull/9422/

File: src/test/java/io/pravega/connectors/flink/utils/StreamSourceOperatorTestHarness.java [hunk 11]
```diff
@@ -41,8 +37,9 @@
 
     private final ConcurrentLinkedQueue<Long> triggeredCheckpoints;
 
-    private final List<RecordWriter<SerializationDelegate<StreamRecord<T>>>> recordWriters;
-    private final OperatorChain<T, StreamOperator<T>> operatorChain;
+    private final RecordWriterDelegate recordWriterDelegate;
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/utils/StreamSourceOperatorTestHarness.java]: This is changed due to https://github.com/apache/flink/pull/10029/

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java [hunk 12]
```diff
@@ -314,7 +314,7 @@ public long extractTimestamp(Integer element, long previousElementTimestamp) {
             expected.add(record(2, 2));
             expected.add(watermark(2));
             // automatic watermark at the end of stream
-            expected.add(watermark(Long.MAX_VALUE));
+            //expected.add(watermark(Long.MAX_VALUE));
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java]: This is because of https://github.com/apache/flink/commit/3171edfe78a7f9feea88488273bf4341eee7b7f0 I think we can remove the line.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 13]
```diff
@@ -310,19 +325,20 @@ private void testTableSourceBatchDescriptor(Stream stream, PravegaConfig pravega
                 .withSchema(schema);
 
         final Map<String, String> propertiesMap = desc.toProperties();
-        final TableSource<?> source = TableFactoryService.find(BatchTableSourceFactory.class, propertiesMap)
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: We should not change the batch test now. We should keep it as is and uncomment the test after fixing #341

File: src/test/java/io/pravega/connectors/flink/utils/StreamSourceOperatorTestHarness.java [hunk 14]
```diff
@@ -52,8 +49,8 @@ public StreamSourceOperatorTestHarness(StreamSource<T, F> operator, int maxParal
         super(operator, maxParallelism, parallelism, subtaskIndex);
         this.sourceOperator = operator;
         this.triggeredCheckpoints = new ConcurrentLinkedQueue<>();
-        this.recordWriters = StreamTask.createRecordWriters(this.config, this.getEnvironment());
-        this.operatorChain = new OperatorChain<>(this.mockTask, recordWriters);
+        this.recordWriterDelegate = StreamTask.createRecordWriterDelegate(this.config, this.getEnvironment());
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/utils/StreamSourceOperatorTestHarness.java]: Actually we don't need the `recordWriterDelegate` as a member of the class, the below sentence is the only place we need that, we can contruct it below.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 15]
```diff
@@ -265,11 +280,15 @@ private void testTableSourceStreamingDescriptor(Stream stream, PravegaConfig pra
                 .createStreamTableSource(propertiesMap);
 
         tableEnv.registerTableSource("MyTableRow", source);
-        String sqlQuery = "SELECT user, count(uri) from MyTableRow GROUP BY user";
+        String sqlQuery = "SELECT user, " +
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: Switched the test SQL in batch and streaming because the batch mode does not support watermark and has some remaining issues on timestamps.

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 16]
```diff
@@ -293,40 +312,36 @@ private void testTableSourceBatchDescriptor(Stream stream, PravegaConfig pravega
 
         // read data from the stream using Table reader
         Schema schema = new Schema()
-                .field("user", Types.STRING)
-                .field("uri", Types.STRING)
-                .field("accessTime", Types.SQL_TIMESTAMP).rowtime(
-                        new Rowtime().timestampsFromField("accessTime")
-                                .watermarksPeriodicBounded(30000L));
+                .field("user", DataTypes.STRING())
+                .field("uri", DataTypes.STRING());
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: Avoid the schema registration with timestamp here due to the known issue #341 .

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 17]
```diff
@@ -358,15 +374,14 @@ static boolean compare(List<Row> results, List<Row> expectedResults) {
...
+                //{ "Tony",  "2018-08-02 10:45:00.0", "1" },
 
+        };
         List<Row> expectedResults = new ArrayList<>();
-
         for (int i = 0; i < data.length; i++) {
             Row row = new Row(3);
             row.setField(0, data[i][0]);
-            row.setField(1, Timestamp.valueOf(data[i][1]));
+            row.setField(1, LocalDateTime.parse(data[i][1], formatter));
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: Blink planner returns `LocalDateTime` instead of the `Timestamp`

File: src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java [hunk 18]
```diff
@@ -358,15 +374,14 @@ static boolean compare(List<Row> results, List<Row> expectedResults) {
                 { "Nina",  "2018-08-02 09:40:00.0", "1" },
                 { "Peter", "2018-08-02 09:45:00.0", "1" },
                 { "Tony",  "2018-08-02 09:45:00.0", "1" },
-                { "Tony",  "2018-08-02 10:45:00.0", "1" }
-        };
+                //{ "Tony",  "2018-08-02 10:45:00.0", "1" },
```
Comment (crazyzhou) [inline src/test/java/io/pravega/connectors/flink/FlinkPravegaTableITCase.java]: The watermark is not passing the `10:45`, so the window will not be fired and compute the count, thus we don't have this line for output. Please put the annotation here.

**Comment by AlexanderZhao1** [comment]
image (1)

Pravega-samples exactly-once test case passed.
